### ISSUE TITLE
restoreCalicoLabels Memory optimization

### ISSUE THREAD

**Issue by gongzixiangyuan** [projectcalico/calico#8627]
## Expected Behavior
In a cluster of 1,000 nodes or more, calico-node/calico-typha consumes as little memory as possible

## Current Behavior
In a 1000-node cluster, kube-controller-manager updates the node timestamp every 5 seconds. For calico-node/calico-typha, this is a continuous change of the node. calico frequently does the following:
1. Store K8S Node’s labels in Calico Node’s annotation (projectcalico.org/kube-labels)
2. Deserialize it from the annotation of Calico Node to avoid overwriting the label of k8s node when the label in calico node is consistent with the k8s node.

## Possible Solution
Why put k8s node's label in annotation and then deserialize it for comparison?
```go
func mergeCalicoNodeIntoK8sNode(calicoNode *libapiv3.Node, k8sNode *kapiv1.Node) (*kapiv1.Node, error) {
    // Can't we get k8sNode directly here?
    calicoNode, err := restoreCalicoLabels(calicoNode)
}
``` 

## Steps to Reproduce (for bugs)
1. View pprof

## Context
I'm trying to deploy a 1000-node calico-node cluster

## Your Environment
* Calico version 3.25.2

**Comment by gongzixiangyuan** [comment]
@fasaxc   @coutinhop  Is it possible to make an enhancement here?

**Comment by gongzixiangyuan** [comment]
In addition, logrus also occupies a certain amount of memory allocation. Some places are debug or trace logs, but logrus.WithField and logrus.WithFields also occupy memory allocation. For example:

```go
func (h *connection) sendMsg(msg interface{}) error {
if h.cxt.Err() != nil {
// Optimisation, don't bother to send if we're being torn down.
return h.cxt.Err()
}
	
h.logCxt.WithField("msg", msg).Trace("Sending message to client")
}
```

**Comment by caseydavenport** [comment]
@gongzixiangyuan I'm afraid I'm not following particularly well - could you provide a bit more detail and perhaps links to the specific lines that you think are causing performance issues?

**Comment by gongzixiangyuan** [comment]
> @gongzixiangyuan I'm afraid I'm not following particularly well - could you provide a bit more detail and perhaps links to the specific lines that you think are causing performance issues?

https://github.com/projectcalico/calico/blob/95fed7d941b2448ec2c80ff46b8e493e243aa383/libcalico-go/lib/backend/k8s/resources/node.go#L465

- Here we serialize the label of K8S Node and cache it.

https://github.com/projectcalico/calico/blob/95fed7d941b2448ec2c80ff46b8e493e243aa383/libcalico-go/lib/backend/k8s/resources/node.go#L350

- Then use "projectcalico.org/kube-labels" to take out the cached labels and deserialize them.

https://github.com/projectcalico/calico/blob/95fed7d941b2448ec2c80ff46b8e493e243aa383/libcalico-go/lib/backend/k8s/resources/node.go#L345C60-L345C68

- Why don't we directly get the labels from the k8sNode parameter?

**Comment by fasaxc** [comment]
@caseydavenport I think @gongzixiangyuan is just saying that we generate more garbage than we could when processing these updates.  Unless I'm misunderstanding, there's no _leak_ here, it's just that we allocate more than you'd like?  Since each Node update replaces the previous one, the garbage collector should take care of it at a CPU cost.  Is Typha actually using a lot of CPU?  If you use the memory pprof it'll show you the total number of allocs at that place in the code but Typha is a long-running component so that number will be huge even though the data is then GCed a few seconds later.

I haven't seen node resources churn that much recently, are you using a recent version of k8s?  Have you explored tuning the component generating the churn?  This used to be a big issue before we added Typha but I think k8s has improved a lot since then (and I think there were always tunables for changing how often kubelet and kub-controller-manager check in; those needed to be tuned for 1k+ nodes).

Note that v3.27 has s lot of Typha improvements in it.  It now uses a lot less CPU, especially in a large cluster and especially if many clients restart at once.

**Comment by gongzixiangyuan** [comment]
@fasaxc  Yes, that's right! The k8s version is 1.28, my cluster has 1k nodes. The node status changes very frequently, so I saw the label-related code in pprof. The k8s component is not tuned, and the parameter values ​​are all default.

List-pod (about 15k pods) and list-node have memory peaks. In addition, the connection read-write buffer and encode/decode buffer of processes such as felix, confd, allocate-tunnel-addrs, etc. The memory consumed by typha is also very large, I want to reduce the memory consumption of typha as much as possible

**Comment by gongzixiangyuan** [comment]
@fasaxc  Just a few more questions

1. k8s already supports filtering hostNetwork on the server side. Will calico consider filtering host network pods on the server side when querying pods, instead of filtering on the client side?

2. The managedFields generated by the ServerSideApply feature of k8s almost double the size of the pod. Is there any good way to reduce the memory peak of the list-pod when calico-node/calico-typha starts?
